actors/builtin/miner/miner_actor.go

```diff
@@ -481,6 +473,177 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
+				}
+
+				// Record sectors for proof verification
+				allSectors = append(allSectors, partitionSnapshot.Sectors)
+				allIgnored = append(allIgnored, partitionSnapshot.Faults)
+				allIgnored = append(allIgnored, partitionSnapshot.Terminated)
+				allIgnored = append(allIgnored, partitionSnapshot.Unproven)
+
+				// Record active sectors for marking faults.
+				active, err := partitionSnapshot.ActiveSectors()
+				if err != nil {
+					return err
+				}
+				err = disputedSectors.Add(partIdx, active)
+				if err != nil {
+					return err
+				}
+
+				// Record disputed power for penalties.
+				//
+				// NOTE: This also includes power that was
+				// activated at the end of the last challenge
+				// window, and power from sectors that have since
+				// expired.
+				disputedPower = disputedPower.Add(partitionSnapshot.ActivePower())
+				return nil
+			})
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partitions")
+
+			allSectorsNos, err := bitfield.MultiMerge(allSectors...)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to merge sector bitfields")
+
+			allIgnoredNos, err := bitfield.MultiMerge(allIgnored...)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to merge fault bitfields")
+
+			// Load sectors.
+			sectors, err := LoadSectors(store, st.Sectors)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors array")
+			sectorInfos, err := sectors.LoadForProof(allSectorsNos, allIgnoredNos)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors to dispute window post")
+
+			// Find the proving period start for the deadline in question.
+			ppStart := st.ProvingPeriodStart
+			if st.CurrentDeadline < params.Deadline {
+				ppStart -= WPoStProvingPeriod
+			}
+
+			targetDeadline := NewDeadlineInfo(ppStart, params.Deadline, currEpoch)
+
+			// Fails if validation succeeds.
+			err = verifyWindowedPost(rt, targetDeadline.Challenge, sectorInfos, post.Proofs)
+			if err == nil {
+				rt.Abortf(exitcode.ErrIllegalArgument, "failed to dispute valid post")
+				return
+			}
+			rt.Log(rtt.INFO, "successfully disputed: %s", err)
+
+			// Ok, now we record faults. This always works because
+			// we don't allow compaction/moving sectors during the
+			// challenge window.
+			//
+			// However, some of these sectors may have been
+			// terminated. That's fine, we'll skip them.
+			faultExpirationEpoch := targetDeadline.Last() + FaultMaxAge
+			powerDelta, err = dlCurrent.RecordFaults(store, sectors, info.SectorSize, QuantSpecForDeadline(targetDeadline), faultExpirationEpoch, disputedSectors)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to declare faults")
+
+			// Delete disputed proof so it can't be charged multiple times.
+			err = proofsSnapshot.Delete(params.PoStIndex)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to delete disputed proof")
```

Comment:
I assume this also shifts the array. May be worth mentioning that if 2 posts are disputable, this causes a race condition where a disputer may need to wait for a previous dispute to be processed before submitting the next dispute.

Reply:
Reference: https://github.com/filecoin-project/specs-actors/blob/d5288c22c9f53a514e72260040cde81667a4e791/actors/builtin/miner/deadline_state.go#L785-L790

Reply:
It just makes a hole, nothing is shifted.